src/stream.ts

```diff
@@ -177,246 +188,265 @@ export class WebRTCStream implements Stream {
+
+  /**
+   * Callback to invoke when the stream is closed.
+   */
+  closeCb?: (stream: WebRTCStream) => void
+
+  constructor (opts: StreamInitOpts) {
+    this.channel = opts.channel
+    this.id = this.channel.label
+
+    this.stat = opts.stat
+    switch (this.channel.readyState) {
+      case 'open':
+        this.opened.resolve()
+        break
+
+      case 'closed':
+      case 'closing':
+        this.streamState.state = StreamStates.CLOSED
+        if (this.stat.timeline.close === undefined || this.stat.timeline.close === 0) {
+          this.stat.timeline.close = new Date().getTime()
+        }
+        this.opened.resolve()
+        break
+      case 'connecting':
+        // noop
+        break
+
+      default:
+        unreachableBranch(this.channel.readyState)
+    }
+
+    this.metadata = opts.metadata ?? {}
+
+    // closable sink
+    this.sink = this._sinkFn
+
+    // handle RTCDataChannel events
+    this.channel.onopen = (_evt) => {
+      this.stat.timeline.open = new Date().getTime()
+      this.opened.resolve()
+    }
+
+    this.channel.onclose = (_evt) => {
+      this.close()
+    }
+
+    this.channel.onerror = (evt) => {
+      const err = (evt as RTCErrorEvent).error
+      this.abort(err)
+    }
+
+    const self = this
+
+    // reader pipe
+    this.channel.onmessage = async ({ data }) => {
+      if (data === null || data.length === 0) {
+        return
+      }
+      this._innersrc.push(new Uint8Array(data as ArrayBufferLike))
+    }
+
+    // pipe framed protobuf messages through a length prefixed decoder, and
+    // surface data from the `Message.message` field through a source.
+    this._src = pipe(
+      this._innersrc,
+      lengthPrefixed.decode(),
+      (source) => (async function * () {
+        for await (const buf of source) {
+          const message = self.processIncomingProtobuf(buf.subarray())
+          if (message != null) {
+            yield new Uint8ArrayList(message)
+          }
+        }
+      })()
+    )
+  }
+
+  // If user attempts to set a new source this should be a noop
+  set source (_src: Source<Uint8ArrayList>) { }
+
+  get source (): Source<Uint8ArrayList> {
+    return this._src
+  }
+
+  /**
+   * Closable sink
+   */
+  private async _sinkFn (src: Source<Uint8ArrayList | Uint8Array>): Promise<void> {
+    await this.opened.promise
+
+    if (this.streamState.isWriteClosed()) {
+      return
+    }
+
+    const self = this
+    const closeWriteIterable = {
+      async * [Symbol.asyncIterator] () {
+        await self.closeWritePromise.promise
+        yield new Uint8Array(0)
+      }
+    }
+
+    for await (const buf of merge(closeWriteIterable, src)) {
+      if (self.streamState.isWriteClosed()) {
+        return
+      }
+
+      const msgbuf = pb.Message.toBinary({ message: buf.subarray() })
+      const sendbuf = lengthPrefixed.encode.single(msgbuf)
+
+      this.channel.send(sendbuf.subarray())
+    }
+
+    // Close the channel after the source has ended
```

Comment:
This tripped me up as well. See https://github.com/libp2p/js-libp2p-webtransport/pull/23

Reply:
The referenced PR and the current implementation would both break on cases similar to:
```
const response = await pipe([request], stream, sink)
// process response
await pipe([request1], stream)
```
However, it might be better to use it-pushable or it-handshake in these cases.
I've refactored the sink to be more readable, and also limited `sink` to being called once similar to the referenced PR.

Reply:
> The referenced PR and the current implementation would both break on cases similar to...

Yes. That's intentional in the js-libp2p design. And the workaround is using a pushable like you mentioned. I'm not a fan of this, but this is how everything else is setup, so it would be even more confusing if this deviated from it.